### PR TITLE
#4: include example folder in gemspec.

### PR DESCRIPTION
**Pull request by fwolfst**
Include examples as requested in #4

### CODE REVIEW

**Comment by lzap** [comment]
Thanks. Can you pull examples/memory.rb file change off the patch?

**Comment by lzap** [comment]
Ping? I'd also prefer explicit list rather than calling git.

**Comment by fwolfst** [comment]
Thanks for remembering me. Should look fine now. What would've been the cleanest way of doing this (I used `git checkout`)?

Also, I do not understand your request about the "explicit list".

**Comment by lzap** [comment]
Something like

```
s.files = Dir["{app,extra,config,db,lib}/**/*"] + Dir["locale/**/*.{po,pot,mo}"] + ["LICENSE", "README.md"]
```

Ruby can do globs and expanding of directories, no need to call a git.

**Comment by fwolfst** [comment]
Ack, I thought you are talking about my git- and github-fu and not about the code itself. Your proposal is more explicit and consistent with current definition, mine more automagic and follows what `bundle gem mynewgem` creates. I'd actually prefer to use

```
spec.files = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
```

(which comes straight off `bundle gem mynewgem`s gemspecs).
I can see the the downsides and benefits of both approaches (white-/blacklisting). Its your project, though.

**Comment by lzap** [comment]
Please list all the benefits and downsides, I only see your approach spawns an extra process and is slower but on the other hand skipping files which are not in git which is a plus that does not buy me. As a Fedora packager, I can tell stories about unwanted build dependencies and building from distribution tarballs. If we can do it without git, let's do it. Much more simple downstream. I deliver this gem in Fedora, EPEL and RHEL today. This is my main concern.

I don't ship examples in the gem because they are low quality, but if you prefer to have them, I am not against. Please rebase your patch without git, I am fine merging that.

TL;DR - I build RPMs from this GEM and git is not available in that context. Your code simply won't work there and I'd need to create explicit patch to fix this.

**Comment by domcleal** [comment]
> TL;DR - I build RPMs from this GEM and git is not available in that context. Your code simply won't work there and I'd need to create explicit patch to fix this.

Using git in the gemspec doesn't affect RPM builds from published gems, we use it in plenty of projects in this org (e.g. foreman_bootdisk) and in Fedora/EPEL.

It's only used when running `gem build` in this repository, so is only needed by the person building the gem and uploading the archive to rubygems.org.  When building RPMs from the published gem, the gemspec contains a full list of files generated when the gem was built.

**Comment by lzap** [comment]
> Using git in the gemspec doesn't affect RPM builds from published gems, we use it in plenty of projects in this org (e.g. foreman_bootdisk) and in Fedora/EPEL.
> 
> It's only used when running `gem build` in this repository, so is only needed by the person building the gem and uploading the archive to rubygems.org.  When building RPMs from the published gem, the gemspec contains a full list of files generated when the gem was built.

That's not true in general, gems which require patching are unpacked and
rebuilt from scratch once again:

https://fedoraproject.org/wiki/Packaging:Ruby#Building_gems

That will definitely not work. Other authors may ignore this, but that's
not an argument :-)

## 

Later,
 Lukas #lzap Zapletal

**Comment by domcleal** [comment]
> That's not true in general, gems which require patching are unpacked and rebuilt from scratch once again: https://fedoraproject.org/wiki/Packaging:Ruby#Building_gems That will definitely not work.

The gem being unpacked is the one from rubygems.org, which contains a gemspec automatically generated during `gem build`.  It isn't the original one stored in this repository.

As an example, foreman_bootdisk's gemspec in the repository uses `git ls-files`: https://github.com/theforeman/foreman_bootdisk/blob/v6.1.0/foreman_bootdisk.gemspec#L25

<pre>
$ gem spec foreman_bootdisk-6.1.0.gem -l --ruby
# -*- encoding: utf-8 -*-
# stub: foreman_bootdisk 6.1.0 ruby lib
Gem::Specification.new do |s|
  s.name = "foreman_bootdisk"
  s.version = "6.1.0"
  s.required_rubygems_version = Gem::Requirement.new(">= 0") if s.respond_to? :required_rubygems_version=
  s.require_paths = ["lib"]
  s.authors = ["Dominic Cleal"]
  s.date = "2016-01-21"
  s.description = "Plugin for Foreman that creates iPXE-based boot disks to provision hosts without the need for PXE infrastructure."
  s.email = "dcleal@redhat.com"
  s.extra_rdoc_files = ["CHANGES.md", "LICENSE", "README.md"]
  s.files = [".tx/config", "CHANGES.md", "LICENSE", "README.md", "app/controllers/concerns/foreman_bootdisk/unattended_controller_ext.rb", "app/controllers/foreman_bootdisk/api/v2/disks_controller.rb", "app/controllers/foreman_bootdisk/api/v2/subnet_disks_controller.rb", "app/controllers/foreman_bootdisk/disks_controller.rb", "app/helpers/concerns/foreman_bootdisk/hosts_helper_ext.rb", "app/models/concerns/foreman_bootdisk/host_ext.rb", "app/models/setting/bootdisk.rb", "app/services/foreman_bootdisk/iso_generator.rb", "app/services/foreman_bootdisk/renderer.rb", "app/views/foreman_bootdisk/disks/help.html.erb", "app/views/foreman_bootdisk/generic_host.erb", "app/views/foreman_bootdisk/host.erb", "config/routes.rb", "config/routes/mount_engine.rb", "db/migrate/20130914211030_create_host_bootdisk_template.rb", "db/migrate/20130915104500_edit_host_bootdisk_template_multinic.rb", "db/migrate/20130915133321_create_kickstart_bootdisk_template.rb", "db/migrate/20130915201457_create_generic_host_bootdisk_template.rb", "db/migrate/20131021095100_edit_host_bootdisk_template_dns_secondary.rb", "db/migrate/20140522185700_change_templatekind_to_bootdisk.rb", "db/seeds.d/50-bootdisk_templates.rb", "lib/foreman_bootdisk.rb", "lib/foreman_bootdisk/engine.rb", "lib/foreman_bootdisk/version.rb", "lib/tasks/bootdisk.rake", "locale/Makefile", "locale/ca/LC_MESSAGES/foreman_bootdisk.mo", "locale/ca/foreman_bootdisk.po", "locale/de/LC_MESSAGES/foreman_bootdisk.mo", "locale/de/foreman_bootdisk.po", "locale/en/LC_MESSAGES/foreman_bootdisk.mo", "locale/en/foreman_bootdisk.po", "locale/en_GB/LC_MESSAGES/foreman_bootdisk.mo", "locale/en_GB/foreman_bootdisk.po", "locale/es/LC_MESSAGES/foreman_bootdisk.mo", "locale/es/foreman_bootdisk.po", "locale/foreman_bootdisk.pot", "locale/fr/LC_MESSAGES/foreman_bootdisk.mo", "locale/fr/foreman_bootdisk.po", "locale/gemspec.rb", "locale/it/LC_MESSAGES/foreman_bootdisk.mo", "locale/it/foreman_bootdisk.po", "locale/ja/LC_MESSAGES/foreman_bootdisk.mo", "locale/ja/foreman_bootdisk.po", "locale/ko/LC_MESSAGES/foreman_bootdisk.mo", "locale/ko/foreman_bootdisk.po", "locale/pt_BR/LC_MESSAGES/foreman_bootdisk.mo", "locale/pt_BR/foreman_bootdisk.po", "locale/ru/LC_MESSAGES/foreman_bootdisk.mo", "locale/ru/foreman_bootdisk.po", "locale/sv_SE/LC_MESSAGES/foreman_bootdisk.mo", "locale/sv_SE/foreman_bootdisk.po", "locale/zanata.xml", "locale/zh_CN/LC_MESSAGES/foreman_bootdisk.mo", "locale/zh_CN/foreman_bootdisk.po", "locale/zh_TW/LC_MESSAGES/foreman_bootdisk.mo", "locale/zh_TW/foreman_bootdisk.po", "test/functional/foreman_bootdisk/api/v2/disks_controller_test.rb", "test/functional/foreman_bootdisk/api/v2/subnet_disks_controller_test.rb", "test/functional/foreman_bootdisk/disks_controller_test.rb", "test/test_plugin_helper.rb"]
  s.homepage = "http://github.com/theforeman/foreman_bootdisk"
  s.licenses = ["GPL-3"]
  s.rubygems_version = "2.2.5"
  s.summary = "Create boot disks to provision hosts with Foreman"
end
</pre>

**Comment by lzap** [comment]
Oh I haven't realized that. I have no objections then! Thanks.

**Comment by lzap** [comment]
Merged as 6e9881e87c0ad45470b3ac446842f2b9d262934c, thank you!

**Comment by fwolfst** [comment]
About pros and cons of the `git ls`-approach:
_Pro_
- common ideom (see `bundle new`)
- fire and forget. Given a development workflow that embraces git, the gem will never miss a file because it was not added to the separate and explicit "manifest". You'll rarely have to modify that line again.

_Con_
- fire and forget. You'll get used to the magic. One day it might haunt you and you might not immediately see what's going on. And, if unknown, definately harder to grok.
- involves seperate process (and you need an actual clone of the repo) when building the gem.
- when not at home with git, .gitignore might make you believe that everything you need is already added to the repo.
- You might accidentally forget to blacklist a file or directory. It eats away diskspace but probably won't hurt otherwise.

Pros and Cons of the explicit manifest:
_Pro_
- You really know whats going on
- building gem does not depend on git repository or tooling

_Con_
- forgot to add a file to the manifest? Build again. This would definately happen more often to _me_ than forgetting to `git add` a file.
- With bad globbing: Oops, had a .swp file flying around? Now its on rubygems.org!

My stance that its up to the maintainer. As I usually start out with `bundle new`, I stick to that.

And in my oppinion most examples are better than no example. And thanks for your efforts!
